v2/rest/client.go

```diff
@@ -0,0 +1,310 @@
+package gohierarchyclient
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/SKF/go-hierarchy-client/v2/rest/models"
+
+	rest "github.com/SKF/go-rest-utility/client"
+	"github.com/SKF/go-utility/v2/stages"
+	"github.com/SKF/go-utility/v2/uuid"
+)
+
+type TreeFilter struct {
+	Depth         int
+	Limit         int
+	Offset        int
+	MetadataKey   string
+	MetadataValue string
+	NodeTypes     []string
+	ModifiedAfter string
+}
+
+type HierarchyClient interface {
+	GetNode(ctx context.Context, id uuid.UUID) (models.GetNodeResponse, error)
+	CreateNode(ctx context.Context, node models.CreateNodeRequest) (models.CreateNodeResponse, error)
+	UpdateNode(ctx context.Context, id uuid.UUID, node models.UpdateNodeRequest) (models.UpdateNodeResponse, error)
+	DeleteNode(ctx context.Context, id uuid.UUID) error
+	DuplicateNode(ctx context.Context, source uuid.UUID, destination uuid.UUID, suffix string) (models.DuplicateNodeResponse, error)
+
+	GetAncestors(ctx context.Context, id uuid.UUID, height int, nodeTypes ...string) (models.GetAncestorsResponse, error)
+	GetCompany(ctx context.Context, id uuid.UUID) (models.GetCompanyResponse, error)
+	GetSubtree(ctx context.Context, id uuid.UUID, filter TreeFilter, continuationToken string) (models.GetSubtreeResponse, error)
+	GetSubtreeCount(ctx context.Context, id uuid.UUID, nodeTypes ...string) (models.GetSubtreeCountResponse, error)
+
+	LockNode(ctx context.Context, id uuid.UUID, recursive bool) error
+	UnlockNode(ctx context.Context, id uuid.UUID, recursive bool) error
+
+	GetOrigins(ctx context.Context, provider, continuationToken string, limit int) (models.GetOriginsResponse, error)
+	GetOriginsByType(ctx context.Context, provider, originType, continuationToken string, limit int) (models.GetOriginsResponse, error)
+	GetProviderNodeIDs(ctx context.Context, provider, continuationToken string, limit int) (models.GetNodesByPartialOriginResponse, error)
+	GetProviderNodeIDsByType(ctx context.Context, provider, originType, continuationToken string, limit int) (models.GetNodesByPartialOriginResponse, error)
+	GetOriginNodeID(ctx context.Context, origin models.Origin) (models.GetNodeByOriginResponse, error)
+}
+
+type client struct {
+	*rest.Client
+	clientID string
+}
+
+func WithStage(stage string) rest.Option {
+	if stage == stages.StageProd {
+		return rest.WithBaseURL("https://api.hierarchy.enlight.skf.com")
+	}
+
+	return rest.WithBaseURL(fmt.Sprintf("https://api.%s.hierarchy.enlight.skf.com", stage))
+}
+
+func NewClient(clientID string, opts ...rest.Option) HierarchyClient {
```

Comment:
To follow the pattern we have with the clients, should the client id be an Option?

Reply:
I thought since the API is unusable without the client ID it was good to give it more visibility than an option,

Reply:
It is also unusable without an authentication token, which is an Option today.

The reason I put it as an option is that it becomes more verbose and clear what the value is for.